File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/BlackjackCard.java
package com.panopset.blackjackEngine;

public class BlackjackCard {

  private final Card c;
  
  public Card getCard() {
    return c;
  }

  public BlackjackCard(CardDefinition cardDefinition) {
    this(new Card(cardDefinition, true), true);
  }

  public BlackjackCard(Card card) {
    this(card, true);
  }

  public BlackjackCard(Card card, boolean isShowing) {
    c = card;
    if (isShowing) {
      c.show();
      count();
    }
  }

  public BlackjackCard(BlackjackCard card) {
    this(card.getCard(), card.getCard().isShowing());
  }

  public boolean isAce() {
    return c.getFace().getOffset() == 0;
  }

  public int getSoftValue() {
    if (isAce()) {
      return 11;
    }
    return getNonAceValue();
  }

  public int getHardValue() {
    if (isAce()) {
      return 1;
    }
    return getNonAceValue();
  }

  private int getNonAceValue() {
    if (c.getFace().getOffset() < 10) {
      return c.getFace().getOffset() + 1;
    } else {
      return 10;
    }
  }

  private void count() {
    CardCounterImpl.count(this);
  }

  public void show() {
    count();
    c.show();
  }

  @Override
  public String toString() {
    return c.name();
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/BlackjackCard.kt
package com.panopset.blackjackEngine

class BlackjackCard(val card: Card) {

    constructor(card: Card, isShowing: Boolean) : this(card) {
        card.isShowing = isShowing
    }

    constructor(cardDefinition: CardDefinition) : this(Card(cardDefinition))
    constructor(blackjackCard: BlackjackCard) : this(blackjackCard.card, blackjackCard.card.isShowing)

    fun isAce(): Boolean {
        return card.face.offset == 0
    }

    fun getSoftValue(): Int {
        return if (isAce()) {
            11
        } else getNonAceValue()
    }

    fun getHardValue(): Int {
        return if (isAce()) {
            1
        } else getNonAceValue()
    }

    private fun getNonAceValue(): Int {
        return if (card.face.offset < 10) {
            card.face.offset + 1
        } else {
            10
        }
    }

    override fun toString(): String {
        return card.name()
    }

    fun show() {
        card.isShowing = true
    }
}
